pkg/storegateway/bucket_stores.go

```diff
@@ -445,7 +445,7 @@ func (t timeoutGate) Start(ctx context.Context) error {
 	defer cancel()
 
 	err := t.delegate.Start(ctx)
-	if errors.Is(context.Cause(ctx), errGateTimeout) {
+	if err != nil && errors.Is(context.Cause(ctx), errGateTimeout) {
```

Comment:
Good job spotting this race! It's very nasty! Even if you have a test, I think a comment here to explain why we check `err != nil` would be very nice and useful for the future. Fancy to open a follow up PR with it?

Reply:
I can add a comment to the `timeoutGate` in #8138

Reply:
Thanks @56quarters!